subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -231,9 +231,9 @@ protected GroovyCompilerFactory getGroovyCompilerFactory() {
     private RecompilationSpecProvider createRecompilationSpecProvider(InputChanges inputChanges, Multimap<String, String> sourceClassesMapping) {
         return new GroovyRecompilationSpecProvider(
             ((ProjectInternal) getProject()).getFileOperations(),
-            getSource(),
+            getStableSources().getAsFileTree(),
             inputChanges.isIncremental(),
-            () -> inputChanges.getFileChanges(getStableSources()),
+            () -> inputChanges.getFileChanges(getStableSources()).iterator(),
```

Comment:
I think we should call `getStableSources()` once.

Reply:
Done.